metadata/src/main/java/org/apache/kafka/image/loader/MetadataBatchLoader.java

```diff
@@ -0,0 +1,257 @@
+        MetadataUpdater callback
+    ) {
+        this.log = logContext.logger(MetadataBatchLoader.class);
+        this.time = time;
+        this.faultHandler = faultHandler;
+        this.leaderAndEpochSupplier = leaderAndEpochSupplier;
+        this.callback = callback;
+    }
+
+    /**
+     * Reset the state of this batch loader to the given image. Any un-flushed state will be
+     * discarded.
+     *
+     * @param image     Metadata image to reset this batch loader's state to.
+     */
+    public void resetToImage(MetadataImage image) {
+        this.image = image;
+        this.delta = new MetadataDelta.Builder().setImage(image).build();
+        this.transactionState = TransactionState.NO_TRANSACTION;
+        this.lastOffset = image.provenance().lastContainedOffset();
+        this.lastEpoch = image.provenance().lastContainedEpoch();
+        this.lastContainedLogTimeMs = image.provenance().lastContainedLogTimeMs();
+        this.numBytes = 0;
+        this.numBatches = 0;
+        this.totalBatchElapsedNs = 0;
+    }
+
+    /**
+     * Load a batch of records from the log. We have to do some bookkeeping here to
+     * translate between batch offsets and record offsets, and track the number of bytes we
+     * have read. Additionally, there is the chance that one of the records is a metadata
+     * version change which needs to be handled differently.
+     * </p>
+     * If this batch starts a transaction, any records preceding the transaction in this
+     * batch will be implicitly added to the transaction.
+     *
+     * @param batch    The reader which yields the batches.
+     * @return         The time in nanoseconds that elapsed while loading this batch
+     */
+
+    public long loadBatch(Batch<ApiMessageAndVersion> batch) {
+        long startNs = time.nanoseconds();
+        int indexWithinBatch = 0;
+
+        lastContainedLogTimeMs = batch.appendTimestamp();
+        lastEpoch = batch.epoch();
+
+        for (ApiMessageAndVersion record : batch.records()) {
+            try {
+                replay(record);
+            } catch (Throwable e) {
+                faultHandler.handleFault("Error loading metadata log record from offset " +
+                    batch.baseOffset() + indexWithinBatch, e);
+            }
+
+            // Emit the accumulated delta if a new transaction has been started and one of the following is true
+            // 1) this is not the first record in this batch
+            // 2) this is not the first batch since last emitting a delta
+            if (transactionState == TransactionState.STARTED_TRANSACTION && (indexWithinBatch > 0 || numBatches > 0)) {
+                MetadataProvenance provenance = new MetadataProvenance(lastOffset, lastEpoch, lastContainedLogTimeMs);
+                LogDeltaManifest manifest = LogDeltaManifest.newBuilder()
+                    .provenance(provenance)
+                    .leaderAndEpoch(leaderAndEpochSupplier.get())
```

Comment:
I don't think we need the `leaderAndEpoch` supplier here since we can just pass in `MetadataLoader.currentLeaderAndEpoch` to this function. The metadata loader already tracks this and it's cheaper than taking locks and contacting the raft layer